src/storage/mod.rs

```diff
@@ -585,6 +587,72 @@ pub(crate) fn source_archive_path(name: &str, version: &str) -> String {
     format!("sources/{name}/{version}.zip")
 }
 
+#[instrument(skip(storage))]
+fn migrate_one(storage: &Storage, archive_path: &str) -> Result<()> {
+    // this will also download the index if it doesn't exist locally
+    let local_index_filename = storage.get_index_filename(archive_path)?;
+
+    if archive_index::is_sqlite_file(&local_index_filename)? {
+        info!("index already in SQLite format, skipping");
+        return Ok(());
+    }
+
+    info!("converting local index...");
+    let remote_index_path = format!("{}.index", &archive_path);
+    let new_index_temp_path = archive_index::convert_to_sqlite_index(&local_index_filename)?;
+
+    // first upload to S3, ongoing requests will still use the local CBOR index
+    info!("uplading to S3...");
+    // S3 put-object will overwrite the existing index
+    storage.store_one(remote_index_path, std::fs::read(&new_index_temp_path)?)?;
+
+    // move the temporary file into place
+    // This has a race condition when a request is trying to read the index between
+    // the `remove_file` and the `rename`. In this case the handler will then just
+    // unnecessarily download the index again.
+    fs::remove_file(&local_index_filename)?;
+    fs::rename(new_index_temp_path, local_index_filename)?;
```

Comment:
Renaming the file here fails for me because `/tmp` is a different mountpoint to the storage:

```
new_index_temp_path="/tmp/.tmpP9ooeu"
local_index_filename="ignored/cratesfyi-prefix/archive_cache/sources/test-shisho-policy-sdk/0.1.8.zip.index"
error converting source archive index: Invalid cross-device link (os error 18)
```

I got that to work by using a temp dir under the archive-cache path

```patch
diff --git src/storage/archive_index.rs src/storage/archive_index.rs                                                                                                                    index 4ffdc6e7..e0b3a8c5 100644
--- src/storage/archive_index.rs
+++ src/storage/archive_index.rs
@@ -288,13 +288,13 @@ pub(crate) fn find_in_file<P: AsRef<Path>>(
     }
 }

-pub(crate) fn convert_to_sqlite_index<P: AsRef<Path>>(path: P) -> Result<TempPath> {
+pub(crate) fn convert_to_sqlite_index<P: AsRef<Path>>(path: P, tmpdir: impl AsRef<Path>) -> Result<TempPath> {
     let path = path.as_ref();
     let index: Index = { serde_cbor::from_reader(BufReader::new(File::open(path)?))? };

     // write the new index into a temporary file so reads from ongoing requests
     // can continue on the old index until the new one is fully written.
-    let tmp_path = tempfile::NamedTempFile::new()?.into_temp_path();
+    let tmp_path = tempfile::NamedTempFile::new_in(tmpdir)?.into_temp_path();
     index
         .write_sqlite(&tmp_path)
         .context("error writing SQLite index")?;
diff --git src/storage/mod.rs src/storage/mod.rs
index 087b7997..2acbe4df 100644
--- src/storage/mod.rs
+++ src/storage/mod.rs
@@ -588,7 +588,7 @@ pub(crate) fn source_archive_path(name: &str, version: &str) -> String {
 }
 #[instrument(skip(storage))]
-fn migrate_one(storage: &Storage, archive_path: &str) -> Result<()> {
+fn migrate_one(storage: &Storage, archive_path: &str, tmpdir: &Path) -> Result<()> {
     // this will also download the index if it doesn't exist locally
     let local_index_filename = storage.get_index_filename(archive_path)?;

@@ -599,7 +599,7 @@ fn migrate_one(storage: &Storage, archive_path: &str) -> Result<()> {

     info!("converting local index...");
     let remote_index_path = format!("{}.index", &archive_path);
-    let new_index_temp_path = archive_index::convert_to_sqlite_index(&local_index_filename)?;
+    let new_index_temp_path = archive_index::convert_to_sqlite_index(&local_index_filename, tmpdir)?;

     // first upload to S3, ongoing requests will still use the local CBOR index
     info!("uplading to S3...");
@@ -621,6 +621,11 @@ pub fn migrate_old_archive_indexes(
     storage: &Storage,
     conn: &mut impl postgres::GenericClient,
 ) -> Result<()> {
+    let tmpdir = storage.config.local_archive_cache_path.join("tmp");
+    if !tmpdir.exists() {
+        fs::create_dir(&tmpdir)?;
+    }
+
     for row in conn
         .query_raw(
             "
@@ -643,10 +648,10 @@ pub fn migrate_old_archive_indexes(
         let version: &str = row.get(1);
         info!("converting archive index for {} {}...", name, version);

-        if let Err(err) = migrate_one(storage, &rustdoc_archive_path(name, version)) {
+        if let Err(err) = migrate_one(storage, &rustdoc_archive_path(name, version), &tmpdir) {
             error!("error converting rustdoc archive index: {:?}", err);
         }
-        if let Err(err) = migrate_one(storage, &source_archive_path(name, version)) {
+        if let Err(err) = migrate_one(storage, &source_archive_path(name, version), &tmpdir) {
             error!("error converting source archive index: {:?}", err);
         }
     }
```

After that it successfully migrated everything on my server.

Reply:
This is awesome! 

I applied your changes, but I chose another base directory. 

From what I see `local_archive_cache_path.join("tmp")` would overlap with the local index storage for the `tmp` crate. 

I assume that everything below `config.prefix` will be on one filesystem, so having the temp-dir there should be fine?

Reply:
`local_archive_cache_path` is sub-divided into the `sources` and `rustdoc` directories, so it just adds a sibling to those, but yeah anywhere under prefix should be fine.